File: leetcode/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.java
import java.util.*;

public class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
     this.val = val;
     this.left = left;
     this.right = right;
     }
 }
 
class Solution {
        int idx;
        public TreeNode BST(int[] preorder, int low, int high){
            if(idx==preorder.length)return null;
            int val= preorder[idx];
            if(val<low || val>high)return null;
            idx++;
            TreeNode nval= new TreeNode(val);
            nval.left=BST(preorder,low,val);
            nval.right=BST(preorder,val,high);
            return nval;
        } 
    public TreeNode bstFromPreorder(int[] preorder) {
        return BST(preorder, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    public static void main(String []args){
        Solution s = new Solution();
        int[] preorder = {8, 5, 1, 7, 10, 12};
        TreeNode root = s.bstFromPreorder(preorder);
    }
}

File: leetcode/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
 
class Solution {
public:
int idx;
    TreeNode*  BST(vector<int> &preorder, int high, int low){
        if(idx==preorder.size())return nullptr;
        int val= preorder[idx];
        if(val<low|| val>high)return nullptr;
        idx++;
        TreeNode* root= new TreeNode(val);
        root->left=BST(preorder, val, low);  
        root->right=BST(preorder, high, val);  
        return root;
    }
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return BST(preorder,INT_MAX, INT_MIN);
    }
};
int main(){
    Solution s;
    vector<int> preorder = {8, 5, 1, 7, 10, 12};
    TreeNode* root = s.bstFromPreorder(preorder);

    // Function to print the inorder traversal of the BST
    function<void(TreeNode*)> inorderTraversal = [&](TreeNode* node) {
        if (node == nullptr) return;
        inorderTraversal(node->left);
        cout << node->val << " ";
        inorderTraversal(node->right);
    };

    cout << "Inorder Traversal of the constructed BST: ";
    inorderTraversal(root);
    cout << endl;

    return 0;
}